File: frontend/src/app/category.ts
export interface Category {
  id: string;
  name: string;
  isMain: boolean,
  thumbnailUrl: string;
  parentCategoryId?: string;
  subcategories: Subcategory[];
}

export interface Subcategory  {
  id: string;
  name: string;
  subcategories?: Subcategory[];
}

File: product-service/src/main/java/org/productservice/entity/Category.java
package org.productservice.entity;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import java.util.ArrayList;
import java.util.List;

@Document(value = "category")
@AllArgsConstructor
@NoArgsConstructor
@Builder
@Data
public class Category {

    @Id
    private String id;

    private String name;

    private String thumbnailUrl;

    private boolean isMain;

    private Category parentCategory;

    private List<Category> subcategories = new ArrayList<>();
}
